be1-go/channel/generalChirping/mod.go

```diff
@@ -0,0 +1,249 @@
+		Msg("received a catchup")
+	return c.inbox.GetSortedMessages()
+}
+
+// broadcastToAllClients is a helper message to broadcast a message to all
+// subscribers.
+func (c *Channel) broadcastToAllClients(msg message.Message) {
+
+	c.log.Info().
+		Str(msgID, msg.MessageID).
+		Msg("broadcast new chirp to all clients")
+
+	rpcMessage := method.Broadcast{
+		Base: query.Base{
+			JSONRPCBase: jsonrpc.JSONRPCBase{
+				JSONRPC: "2.0",
+			},
+			Method: query.MethodBroadcast,
+		},
+		Params: struct {
+			Channel string          `json:"channel"`
+			Message message.Message `json:"message"`
+		}{
+			c.channelPath,
+			msg,
+		},
+	}
+
+	buf, err := json.Marshal(&rpcMessage)
+	if err != nil {
+		c.log.Err(err).Msg("failed to marshal broadcast query")
+	}
+
+	c.sockets.SendToAll(buf)
+}
+
+// VerifyBroadcastMessage checks if a Broadcast message is valid
+func (c *Channel) VerifyBroadcastMessage(broadcast method.Broadcast) error {
+	c.log.Info().Msg("received broadcast")
+
+	// Check if the structure of the message is correct
+	msg := broadcast.Params.Message
+
+	jsonData, err := base64.URLEncoding.DecodeString(msg.Data)
+	if err != nil {
+		return xerrors.Errorf("failed to decode message data: %v", err)
+	}
+
+	// Verify the data
+	err = c.hub.GetSchemaValidator().VerifyJSON(jsonData, validation.Data)
+	if err != nil {
+		return xerrors.Errorf("failed to verify json schema: %w", err)
+	}
+
+	// Check if the message already exists
+	if _, ok := c.inbox.GetMessage(msg.MessageID); ok {
+		return answer.NewError(-3, "message already exists")
+	}
+
+	return nil
+}
+
+// AddChirp checks and stores an add chirp message
+func (c *Channel) AddChirp(msg message.Message) error {
+	err := c.verifyChirpBroadcastMessage(msg)
+	if err != nil {
+		return xerrors.Errorf("failed to get and verify add chirp message: %v", err)
+	}
+	c.inbox.StoreMessage(msg)
+
+	return nil
+}
+
+// DeleteChirp checks and stores a delete chirp message
+func (c *Channel) DeleteChirp(msg message.Message) error {
+	err := c.verifyChirpBroadcastMessage(msg)
+	if err != nil {
+		return xerrors.Errorf("failed to get and verify delete chirp message: %v", err)
+	}
+	c.inbox.StoreMessage(msg)
+
+	return nil
+}
+
+// verifyChirpBroadcastMessage verify a chirp broadcast message
+func (c *Channel) verifyChirpBroadcastMessage(msg message.Message) error {
+	var chirpMsg messagedata.ChirpBroadcast
+
+	err := msg.UnmarshalData(&chirpMsg)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal cast vote: %v", err)
+	}
+
+	senderBuf, err := base64.URLEncoding.DecodeString(msg.Sender)
+	if err != nil {
+		return xerrors.Errorf("failed to decode sender key: %v", err)
+	}
+
+	senderPoint := crypto.Suite.Point()
+	err = senderPoint.UnmarshalBinary(senderBuf)
+	if err != nil {
+		return answer.NewError(-4, "invalid sender public key")
+	}
+
+	ok := c.hub.GetPubkey().Equal(senderPoint)
+	if !ok {
+		return answer.NewError(-4, "only organizer can broadcast the chirp messages")
```

Comment:
Small nitpick, put a server can be a witness or an organizer, so the message isn't 100% correct